### PR TITLE
Restructure JavaScript exceptions

### PR DESCRIPTION
**Pull request by pepone**
- Exceptions.js the base exception class
- LocalException.js the LocalException base class
- LocalExceptions.js all local exception derived type
- UserException.js the user exception base class

Also update the UserException constructor to not take a cause argument, as per #2527

### CODE REVIEW

File: js/src/Ice/UserException.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+import { Exception } from "./Exception.js";
+
+export class UserException extends Exception {
+    constructor() {
+        super();
+        Exception.captureStackTrace(this);
```
Comment (pepone) [inline js/src/Ice/UserException.js:10]: We can probably remove this, not sure why we were capturing stack traces for user exceptions

**Comment by pepone** [comment]
The only relevant change is the updated UserException, constructor. The rest is to move classes around and update the import paths.

File: js/src/Ice/UserException.js (hunk 1 above)
Comment (bernardnormier) [reply js/src/Ice/UserException.js:10]: We shoud remove it. Can we then remove the constructor entirely?

File: js/src/Ice/LocalExceptions.js [hunk 2]
```diff
@@ -0,0 +1,788 @@
...
+const Ice = {
+    ...Ice_Identity,
+    ...Ice_Version,
+};
+
+/**
+ *  This exception is raised when a failure occurs during initialization.
+ **/
+export class InitializationException extends LocalException {
+    constructor(reason = "", _cause = "") {
```
Comment (bernardnormier) [inline js/src/Ice/LocalExceptions.js:18]: I don't understand why the reason is defaulted to "". We should also provide a reason when we construct an InitializationException.

Is it possible to make a class as final/sealed in JS? If yes, then, we should use it in this file.

File: js/src/Ice/LocalExceptions.js [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+    }
+}
+
+/**
+ *  An attempt was made to register something more than once with the Ice run time. This exception is raised if an
+ *  attempt is made to register a servant, servant locator, facet, value factory, plug-in, object adapter, object, or
+ *  user exception factory more than once for the same ID.
+ **/
+export class AlreadyRegisteredException extends LocalException {
+    constructor(kindOfObject = "", id = "", _cause = "") {
```
Comment (bernardnormier) [inline js/src/Ice/LocalExceptions.js:34]: Same here, kindOfObject and id must always be provided by the caller.

File: js/src/Ice/Exception.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ const toString = function (key, object, objectTable, ident) {
 //
 // Ice.Exception
 //
-class Exception extends Error {
+export class Exception extends Error {
```
Comment (bernardnormier) [inline js/src/Ice/Exception.js:41]: We should use/fill Error.cause and Error.message, and definitely not define our own ice_cause property.

**Review by bernardnormier — APPROVED** [review]
I assume the actual refactoring of the local exceptions is for a follow-up PR?

File: js/src/Ice/LocalExceptions.js (hunk 2 above)
Comment (pepone) [reply js/src/Ice/LocalExceptions.js:18]: I will cleanup in a follow-up PR

**Pull request merged**